tutorials/haskell/servant-auth/tutorial.md

```diff
@@ -0,0 +1,566 @@
+[Stack Builders's tutorials repository on GitHub](https://github.com/stackbuilders/tutorials),
+so you can just clone the repository and start playing with it.
+
+```haskell
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE OverloadedStrings    #-}
+{-# LANGUAGE RecordWildCards      #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Main (main) where
+
+import Control.Monad.Catch (try)
+import Control.Monad.Except
+import Control.Monad.Reader
+import Crypto.Cipher.AES (AES256)
+import Crypto.Cipher.Types (ctrCombine)
+import Crypto.Hash.Algorithms (SHA256)
+import Crypto.Random
+import Data.ByteString (ByteString)
+import Data.Serialize hiding (Get)
+import Data.Text (Text)
+import Lucid
+import Network.URI hiding (scheme)
+import Network.Wai
+import Network.Wai.Handler.Warp
+import Servant
+import Servant.HTML.Lucid
+import Servant.Server.Experimental.Auth
+import Servant.Server.Experimental.Auth.Cookie
+import qualified Data.ByteString    as B
+import qualified Data.Text          as T
+import qualified Data.Text.Encoding as T
+```
+
+## The handler monad
+
+To work with cookies, we will need to be able to access at least three
+values from handlers: `AuthCookieSettings` (more on that below),
+`RandomSource`, and `ServerKey`. For type-safe rendering we will need three
+more things: approot (as a `String`), port (`Int`), and scheme (`String`).
+Knowing these requirements it's an obvious choice to use monad stack with
+`ReaderT` parametrized over `AppContext` record containing all necessary
+data that should be available in handlers:
+
+```haskell
+-- | The handler monad, to work with cookies we need to have access to
+-- 'AuthCookieSettings', 'RandomSource', and 'ServerKey'. It's also a handy
+-- place to put some values that will be useful for type-safe route
+-- rendering.
+
+type App = ReaderT AppContext (ExceptT ServantErr IO)
+
+-- | The application's context.
+
+data AppContext = AppContext
+  { appContextAuthSettings :: AuthCookieSettings
+  , appContextRandomSource :: RandomSource
+  , appContextServerKey    :: ServerKey
+  , appContextApproot      :: String
+  , appContextPort         :: Int
+  , appContextScheme       :: String
+  }
+```
+
+This should be pretty self-explanatory, so let's move on to the session data.
+
+## Session data
+
+First of all it's necessary to decide which data to include in session. It
+will be stored in encrypted form, on client's machine in a cookie, which
+means that every time client makes a request to your domain, all the data
+will be sent to you — this is why we should not store anything big there. In
+most cases an identifier of the logged in user and some basic information
+that forms context of current session is stored. In our example, we will
+store user name and email:
```

Comment:
@mrkkrp 

- "to include in the session"
- "will be stored in encrypted form on the client's machine in a cookie"
- "every time a client makes a request"